rustls/src/compress.rs

```diff
@@ -0,0 +1,77 @@
+//! Certificate compression and decompression support
+
+use alloc::vec::Vec;
+use core::fmt::Debug;
+
+use crate::enums::CertificateCompressionAlgorithm;
+
+/// An available certificate decompression algorithm.
+pub trait CertDecompressor: Debug + Send + Sync {
+    /// Which algorithm this decompressor handles.
+    fn algorithm(&self) -> CertificateCompressionAlgorithm;
+
+    /// Decompress `input`, writing the result to `output`.
+    ///
+    /// `output` matches the declared length of the decompressed data.  The
+    /// decompression process must precisely fill `output`.  If it produces
+    /// more or fewer bytes than fit in `output`, `Err(DecompressionFailed)` must
+    /// be returned.
+    ///
+    /// `Err(DecompressionFailed)` should be returned if the `input` is in any
+    /// way malformed.
```

Comment:
Maybe the sentence about returning this error for the case where the # of bytes produced is a mismatch to `output` size should be rolled into this sentence? Something like:

> `Err(DecompressionFailed)` should be returned if decompression produces more, or fewer bytes than fit in `output`, or if the `input` is in any way malformed.